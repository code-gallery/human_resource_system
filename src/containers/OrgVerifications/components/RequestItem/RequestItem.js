import React, { Component } from 'react'
import PropTypes from 'prop-types'
import Avatar from 'components/Avatar'
import 'containers/UserVerifications/style.css'
import { getDetails } from 'containers/UserVerifications/components/VerificationItem'
import Action from '../Action'

class RequestItem extends Component {
  constructor(props) {
    super(props)
    this.state = {
      showReason: false,
      reasonError: false
    }
  }

  toggleShowReason = () => {
    const { showReason } = this.state
    this.setState({
      showReason: !showReason
    })
  }

  onConfirm = () => {
    const { item } = this.props
    const state = this.state
    const reasons = []
    this.setState({ reasonError: false })

    for (const prop in state) {
      if (prop.indexOf('reason-') !== -1 && state[prop].reason !== '') {
        reasons.push({
          label: state[prop].label,
          reason: state[prop].reason
        })
      }
    }

    if (reasons.length === 0) {
      this.setState({ reasonError: true })
    } else {
      // at least one reason for rejecting claim
      this.props.onReject(item.id, reasons)
    }
  }

  onChangeInput = ({ name, label, reason }) => {
    this.setState({
      [name]: { label, reason }
    })
  }

  render() {
    const { item, onVerify } = this.props
    const { showReason, reasonError } = this.state
    const payload = JSON.parse(item.payload)
    const details = getDetails(payload)
    const name = `${item.user.first_name} ${item.user.last_name}`
    const { type } = item

    return (
      <tr className="VerificationItem">
        <td className="col-xs-12 col-md-3">
          <Avatar
            size="45"
            theme="blue"
            label={name}
            imageUrl={item.user.profile_image}
          />
        </td>
        <td className="col-xs-12 col-md-2 text-center">
          <p className="VerificationItem-type">
            {type}
          </p>
        </td>
        <td className="col-xs-12 col-md-4">
          <ul className="table-list">
            {
              details.map((text, index) => {
                const name = `reason-${index}`
                return (
                  <li className="details" key={index}>
                    <span>{text}</span>
                    {showReason &&
                      <input
                        type="text"
                        placeholder="Suggested change"
                        onChange={(event) => {
                          this.onChangeInput({
                            name,
                            label: text,
                            reason: event.target.value
                          })
                        }}
                      />
                    }
                  </li>
                )
              })
            }
          </ul>
        </td>
        <td className="col-xs-12 col-md-3">
          <Action
            id={item.id}
            showReason={showReason}
            reasonError={reasonError}
            onVerify={onVerify}
            toggle={this.toggleShowReason}
            onConfirm={this.onConfirm}
          />
        </td>
      </tr>
    )
  }
}

RequestItem.propTypes = {
  item: PropTypes.shape({
    payload: PropTypes.string,
    type: PropTypes.string,
    decline_reason: PropTypes.oneOfType([
      PropTypes.object,
      PropTypes.string
    ])
  }),
  onVerify: PropTypes.func.isRequired,
  onReject: PropTypes.func.isRequired
}

export default RequestItem
