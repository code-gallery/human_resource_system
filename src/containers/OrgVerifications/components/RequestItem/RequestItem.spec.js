import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import RequestItem from './RequestItem'

describe('containers/Organisations/OrgVerifications/components/RequestItem/RequestItem', () => {
  const props = {
    onVerify: jest.fn(),
    onReject: jest.fn(),
    item: {
      payload: '{"date":"1994-07-07","start_date":"1994-07-07"}',
      type: 'type',
      id: 111,
      user: {
        first_name: 'Jane',
        last_name: 'Doe',
        profile_image: 'profile_image'
      },
      decline_reason: 'Lorem ipsum dolor'
    }
  }

  describe('when props is valid', () => {
    let wrapper

    beforeEach(() => {
      wrapper = shallow(<RequestItem {...props} />)
    })

    it('renders correctly', () => {
      const tree = renderer.create(
        <RequestItem {...props} />
      ).toJSON()
      expect(tree).toMatchSnapshot()
    })

    it('renders correct Avatar', () => {
      expect(wrapper.find('Avatar').length).toEqual(1)
    })

    it('renders correct Action', () => {
      expect(wrapper.find('Action').length).toEqual(1)
    })
  })
})
