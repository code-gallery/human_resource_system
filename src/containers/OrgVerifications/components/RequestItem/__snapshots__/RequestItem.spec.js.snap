// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/Organisations/OrgVerifications/components/RequestItem/RequestItem when props is valid renders correctly 1`] = `
<tr
  className="VerificationItem"
>
  <td
    className="col-xs-12 col-md-3"
  >
    <div
      className="Avatar Avatar-blue Avatar-size-45 Avatar-withLabel"
    >
      <img
        alt="Jane Doe"
        src="appii-placeholder.png"
      />
      <span>
        Jane Doe
      </span>
    </div>
  </td>
  <td
    className="col-xs-12 col-md-2 text-center"
  >
    <p
      className="VerificationItem-type"
    >
      type
    </p>
  </td>
  <td
    className="col-xs-12 col-md-4"
  >
    <ul
      className="table-list"
    >
      <li
        className="details"
      >
        <span>
          July 1994
        </span>
      </li>
      <li
        className="details"
      >
        <span>
          July 1994
        </span>
      </li>
    </ul>
  </td>
  <td
    className="col-xs-12 col-md-3"
  >
    <div
      className="Action"
    >
      <button
        className="btn green-btn"
        data-verify="true"
        onClick={[Function]}
      >
        Verify this claim
      </button>
      <button
        className="btn red-btn"
        data-reject="true"
        onClick={[Function]}
      >
        Reject this claim
      </button>
    </div>
  </td>
</tr>
`;
